esp-hal/src/i2c.rs

```diff
@@ -261,17 +398,15 @@ where
         bytes: &[u8],
         buffer: &mut [u8],
     ) -> Result<(), Self::Error> {
-        self.peripheral
-            .master_write_read(address, bytes, buffer)
-            .inspect_err(|_| self.internal_recover())
+        self.write_read(address, bytes, buffer)
     }
 }
 
 impl<T, DM: crate::Mode> embedded_hal::i2c::ErrorType for I2C<'_, T, DM> {
     type Error = Error;
 }
 
-impl<T, DM: crate::Mode> embedded_hal::i2c::I2c for I2C<'_, T, DM>
+impl<T> embedded_hal::i2c::I2c for I2C<'_, T, crate::Blocking>
```

Comment:
I'm somewhat hesitant here, is there a reason to explicitly disallow blocking functions on Async APIs?

Reply:
I don't think there is a technical reason and it should be fine

But I also think we are currently already a bit inconsistent how this is handled by different drivers. I _think_ most of the drivers disallow blocking functions in async mode

We should probably decide if we want to change / align that for all drivers. And also, are we fine with an `_async` postfix for the async functions then?

Reply:
We should definitely be more consistent and I have some ideas here - good thing to run through in the next cycle. However, this change does remove a lot of previously available functions, so it's a breaking change, isn't it? It would at least have to be documented.

Reply:
I don't think this is a breaking change currently - it only removes non-public functions if I'm not missing anything but I am open to make this a breaking change and 

- rename the async functions to `*_async`
- implement the blocking functions for any mode

Reply:
I should look at the code blocks more closely next time :) You're right, this is fine for now.